src/Agoda.Analyzers.Test/CodeFixes/AgodaCustom/AG0040LogTemplateAnalyzerUnitTests.cs

```diff
@@ -0,0 +1,216 @@
+            ExpectedFix = "_logger.Information(\"Received request from {Ip} at {DateTime.Now}\", ip, DateTime.Now);",
+            ExpectedDiagnostics = new[]
+            {
+                new DiagnosticResult(AG0040LogTemplateAnalyzer.Rule)
+                    .WithSpan(13, 33, 13, 80)
+                    .WithArguments("string interpolation")
+            }
+        }).SetName("Serilog with string interpolation - complex expression");
+
+        yield return new TestCaseData(new TestCase
+        {
+            Usings = "using Microsoft.Extensions.Logging;\nusing Serilog;",
+            SetupCode =
+                "private readonly Microsoft.Extensions.Logging.ILogger _msLogger;\nprivate readonly Serilog.ILogger _seriLogger;private string info;",
+            LogStatement =
+                "_msLogger.LogInformation($\"MS: {info}\");\n            _seriLogger.Information(\"Seri: \" + info);",
+            ExpectedFix =
+                "_msLogger.LogInformation(\"MS: {Info}\", info);\n            _seriLogger.Information(\"Seri: {Info}\", info);",
+            ExpectedDiagnostics = new[]
+            {
+                new DiagnosticResult(AG0040LogTemplateAnalyzer.Rule)
+                    .WithSpan(15, 38, 15, 51)
+                    .WithArguments("string interpolation"),
+                new DiagnosticResult(AG0040LogTemplateAnalyzer.Rule)
+                    .WithSpan(16, 37, 16, 52)
+                    .WithArguments("string concatenation")
+            }
+        }).SetName("Multiple loggers - both with issues");
+
+        yield return new TestCaseData(new TestCase
+        {
+            Usings = "using Microsoft.Extensions.Logging;",
+            SetupCode = "private readonly ILogger _logger;private string username;private string ipAddress;",
+            LogStatement =
+                "_logger.LogInformation(\"User {Username} logged in from {IpAddress}\", username, ipAddress);",
+            ExpectedFix =
+                "_logger.LogInformation(\"User {Username} logged in from {IpAddress}\", username, ipAddress);",
+            ExpectedDiagnostics = new DiagnosticResult[] { }
+        }).SetName("ILogger with correct usage - multiple parameters with PascalCase");
+
+        yield return new TestCaseData(new TestCase
+        {
+            Usings = "using Serilog;",
+            SetupCode = "private readonly ILogger _logger;",
+            LogStatement = "_logger.Information(\"Status: {@Status}\", new { Code = 200, Message = \"OK\" });",
+            ExpectedFix = "_logger.Information(\"Status: {@Status}\", new { Code = 200, Message = \"OK\" });",
+            ExpectedDiagnostics = new DiagnosticResult[] { }
+        }).SetName("Serilog with correct usage - complex object logging");
+    }
+
+    [Test]
+    [TestCaseSource(nameof(TestCases))]
+    public async Task TestLogTemplateUsage(TestCase testCase)
+    {
+        var test = $@"
+using System;
+{testCase.Usings}
+
+namespace TestNamespace
```

Comment:
Just my 2 cents but I would optimize for readability instead of reduced code duplication in this case. I find the test cases a bit difficult to follow when I need to do string temp laying in my head.

Reply:
I go the other way.
I love to use test cases. Just keep the test function itself really straight-forward and it should be basically:
"Setup", "Input", "expectedOutput" type of tests.


Then you can just read the TestDate and understand all the different tests cases.

Reply:
i agree about test readability, i'll fix it up in another MR, there's 2 like this now, so I'll come up with a better solution and fix them all in one go